torch/csrc/jit/passes/hoist_conv_packed_params.cpp

```diff
@@ -0,0 +1,129 @@
+#include <stack>
+
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/api/module.h>
+#include <torch/csrc/jit/passes/constant_pooling.h>
+#include <torch/csrc/jit/passes/constant_propagation.h>
+#include <torch/csrc/jit/passes/hoist_conv_packed_params.h>
+
+namespace torch {
+namespace jit {
+
+// Hoists packed params from a conv module to the parent module.
+// The benefit is that after this hoisting, the conv module
+// no longer holds anything and can be deleted, reducing model
+// size.
+//
+// Before (easy case):
+//
+// %1 = prim::GetAttr[name="conv1"][%self]
+// %2 = prim::GetAttr[name="_packed_params][%1]
+//
+// After (easy case):
+//
+// %2 = prim::GetAttr[name="{prefix}.conv1._packed_params"][%self]
+//
+// Before (generic case):
+//
+// %1 = prim::GetAttr[name="name1"][%self]
+// %2 = prim::GetAttr[name="name2"][%1]
+// ...
+// %n = prim::GetAttr[name="_packed_params][%n-1]
+//
+// After (generic case):
+//
+// %2 = prim::GetAttr[name="{prefix}.name1{...}.name(n-1)._packed_params"][%self]
+//
+void hoistConvPackedParams(Module& rootModule, Node* getConvPackedParamsNode,
+    const std::string& prefix) {
+
+  auto method = rootModule.get_method("forward");
+  auto graph = method.graph();
+
+  // create a vector of the getAttr nodes
+  //   idx 0 is the node from root to child 1
+  //   idx n-1 is the node from child n-1 to conv
+  std::vector<Node*> getAttrNodes;
+  Node* curNode = getConvPackedParamsNode->inputs()[0]->node();
+  while (!(curNode->outputs()[0]->type() == graph->inputs()[0]->type())) {
+    TORCH_CHECK(
+      curNode->kind() == prim::GetAttr,
+      "Attempted to add a non-prim::GetAttr node to a chain of prim::getAttr nodes.");
+    getAttrNodes.insert(getAttrNodes.begin(), curNode);
+    curNode = curNode->inputs()[0]->node();
+  }
+  TORCH_CHECK(getAttrNodes.size() > 0, "Did not find a chain of prim::getAttr nodes");
```

Comment:
we have some helper function in quantization to do this: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit16getInvokedModuleERNS0_6ModuleEPNS0_4NodeEPNS0_5ValueE, feel free to use them

Reply:
oh this one: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit19getInvokedModuleOptERKNS0_6ModuleEPNS0_4NodeEPNS0_5ValueE since maybe we will fail to get module